sklearn/utils/__init__.py

```diff
@@ -186,20 +189,17 @@ def safe_indexing(X, indices, axis=0):
     ----------
     X : array-like, sparse-matrix, list, pandas.DataFrame, pandas.Series
         Data from which to sample rows, items or columns.
-    indices : array-like
-        - When ``axis=0``, indices need to be an array of integer.
-        - When ``axis=1``, indices can be one of:
-            - scalar: output is 1D, unless `X` is sparse.
-              Supported data types for scalars:
-                - integer: supported for arrays, sparse matrices and
-                  dataframes.
-                - string (key-based): only supported for dataframes.
-            - container: lists, slices, boolean masks: output is 2D.
-              Supported data types for containers:
-                - integer or boolean (positional): supported for
-                  arrays, sparse matrices and dataframes
-                - string (key-based): only supported for dataframes. No keys
-                  other than strings are allowed.
+    indices : bool, int, str, array-like
+        - To select a single element (i.e. row or column), `indices` can be one
+          of the following: `bool` or `int` which are supported by all types of
+          `X`. `indices` being a `str` is only supported for `X` being a
+          dataframe. The selected subset will be 1D, unless `X` is a sparse
+          matrix in which case it will be 2D.
+        - To select multiple elements (i.e. rows or columns), `indices` can be
+          one of the following: `list`, `array`, `slice`. The type used in
+          these containers can be one of the following: `int`, `bool`, and
+          `str`. `str` is only supported when `X` is a dataframe.
```

Comment:
what are slices of bool and str?

Would this work?

 > To select multiple elements (i.e. rows or columns), `indices` can be
          a slice, or a container (`list` or`array`). The type used in
          these containers can be one of the following: `int`, `bool`, and
          `str`. `str` is only supported when `X` is a dataframe.

Reply:
slice of bool will give you the first line (it casting stuff into int). We should raise an error for such case I think.

slice of string is useful with dataframe:

```python
df.loc[slice('xxx', 'yyy')]
```
I see that we have an issue because we only use `iloc` and then it would not work. So the docstring is wrong

Reply:
do we really want to support slices of strings? I'd say we don't for now to keep it simple?

Reply:
It is already supported for the column with dataframe

Reply:
What I mean is:

```python
from sklearn.datasets import fetch_openml
iris = fetch_openml('iris', as_frame=True)
from sklearn.preprocessing import FunctionTransformer
from sklearn.compose import make_column_transformer
transformer = make_column_transformer(
    (FunctionTransformer(), slice('sepallength', 'sepalwidth'))
)
transformer.fit_transform(iris.data, iris.target)
```